RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc

```diff
@@ -168,104 +92,165 @@ void PFClusterEMEnergyCorrector::correctEnergies(const edm::Event &evt, const ed
     reco::PFCluster &cluster = cs[idx];
     
     double e = cluster.energy();
-    double pt = cluster.pt(); 
-    
+    auto const& crep = cluster.position();
+    double pt = cluster.energy()*crep.rho()/crep.r();
+   
     //limit raw energy value used to evaluate corrections
     //to avoid bad extrapolation
     double evale = e;
-    if (_maxPtForMVAEvaluation>0. && pt>_maxPtForMVAEvaluation) {
-      evale *= _maxPtForMVAEvaluation/pt; 
+    if (maxPtForMVAEvaluation_ > 0. && pt > maxPtForMVAEvaluation_) {
+      evale *= maxPtForMVAEvaluation_/pt; 
     }
     
     double invE = (e == 0.) ? 0. : 1./e; //guard against dividing by 0.
     
-    int size = lazyTool.n5x5(cluster);
+    bool iseb = (cluster.layer() == PFLayer::ECAL_BARREL);
+    int clusFlag = 0;
     
-    bool iseb = cluster.layer() == PFLayer::ECAL_BARREL;
+    if (iseb){
+      EBSrFlagCollection::const_iterator srf = ebSrFlags->find(readOutUnitOf(static_cast<EBDetId>(cluster.seed())));      
+      clusFlag = srf->value();
+    } else {
+      EESrFlagCollection::const_iterator srf = eeSrFlags->find(readOutUnitOf(static_cast<EEDetId>(cluster.seed())));      
+      clusFlag = srf->value();
+    }
 
-    //find index of corrections (0-4 for EB, 5-9 for EE, depending on cluster size and raw pt)
-    int coridx = std::min(size,3)-1;
-    if (coridx==2) {
-      if (pt>4.5) {
-        coridx += 1;
-      }
-      if (pt>18.) {
-        coridx += 1;
-      }
+    //find index of corrections (0-3 for EB, 4-7 for EE, depending on cluster size and raw pt)
+    int coridx = 0;
+    int regind = 0;
+
+    if (!iseb) regind = 4;
+
+    if (clusFlag==1)
+      coridx = 0 + regind;
+    else {
+      if (pt<2.5) coridx = 1 + regind;
+      else if (pt>2.5 && pt<6.) coridx = 2 + regind;
+      else if (pt>6.) coridx = 3 + regind;
     }
-    if (!iseb) {
-      coridx += 5;
+    if (clusFlag!=1 && clusFlag!=3) {
+      edm::LogWarning("PFClusterEMEnergyCorrector") << "We can only correct regions readout in ZS (flag 1) or FULL readout (flag 3). Flag " << clusFlag << " is not recognized.";
+      edm::LogWarning("PFClusterEMEnergyCorrector") << "Assuming FULL readout and continuing";
     }
-    
+           
     const GBRForestD &meanforest = *forestH_mean[coridx].product();
     const GBRForestD &sigmaforest = *forestH_sigma[coridx].product();
     
     //find seed crystal indices
-    int ietaix=0;
-    int iphiiy=0;
+    int ietaix = 0;
+    int iphiiy = 0;
+
     if (iseb) {
       EBDetId ebseed(cluster.seed());
       ietaix = ebseed.ieta();
       iphiiy = ebseed.iphi();
-    }
-    else {
+    } else {
       EEDetId eeseed(cluster.seed());
       ietaix = eeseed.ix();
       iphiiy = eeseed.iy();      
     }
 
-    
+    // Form ietamod20 and iphimod20 as well
+    int signeta = (ietaix > 0) ? 1 : -1;
+    int ietamod20 = (std::abs(ietaix) < 26) ? ietaix - signeta : (ietaix - 26*signeta) % 20;
```

Comment:
what are "26" and "20" in this context? Are they available from some header or geometry as constants?